### PR TITLE
feat: Implement zero-shot-image-classification evaluator

### LINKED ISSUE #422: Evaluate zero shot image classification
Zero-shot image classification predicts a label for an image without any task-specific fine-tuning. A dual-encoder model (CLIP-family, SigLIP) produces an image embedding and a set of per-class text embeddings, then picks the class whose text embedding has the highest cosine similarity with the image. The class vocabulary is defined at inference time, not at training time — hence *zero-shot*.

This issue is to track the evaluation work of this kind of model.

### PR DESCRIPTION
**Pull request by zhenchaoni**
# Zero-Shot Image Classification — Design

## 1. What the task is

Zero-shot image classification predicts a label for an image without any task-specific fine-tuning. A dual-encoder model (CLIP-family, SigLIP) produces an image embedding and a set of per-class text embeddings, then picks the class whose text embedding has the highest cosine similarity with the image. The class vocabulary is defined at inference time, not at training time — hence *zero-shot*.

The same model supports any vocabulary. Swap the candidate class list and you have a different classifier; no re-training.

## 2. Split-encoder ONNX export

CLIP/SigLIP are dual-encoder architectures. They fuse only at the cosine-similarity step — the image tower and text tower run independently and do not share activations. We export each as a **separate ONNX file**:

| Role | HF task used for export | Output name |
| --- | --- | --- |
| `image-encoder` | `image-feature-extraction` | `image_embeds` (CLIP WithProjection) / `pooler_output` (SigLIP) |
| `text-encoder`  | `feature-extraction`       | `text_embeds` (CLIP WithProjection) / `pooler_output` (SigLIP) |

Build-time orchestration:

- Composite class: `WinMLModelForZeroShotImageClassification`.
- Registered for `("clip", "zero-shot-image-classification")` and `("siglip", "zero-shot-image-classification")`.
- `forward()` runs the two sub-sessions and calculate the cosine similarity.
- Drop-in compatible with the HF `ZeroShotImageClassificationPipeline`.

## 3. CLI: passing the two ONNX files

`winml eval` accepts the composite model as repeated `-m role=path` pairs:

```bash
winml eval \
  -m image-encoder=path/to/image_encoder.onnx \
  -m text-encoder=path/to/text_encoder.onnx \
  --model-id openai/clip-vit-base-patch16 \
  --task zero-shot-image-classification \
  --device npu \
  --dataset uoft-cs/cifar100 \
  --split test \
  --samples 1000 \
  --column input_column=img \
  --column label_column=fine_label
```

## 4. Evaluation

We leverage OpenAI CLIP's zero-shot classification methodology on CIFAR-100 (`uoft-cs/cifar100`, `test` split). We use classification accuracy as the metric to detect regression.

Steps:

1. **Build prompt list** from the 100 CIFAR-100 class names using HF pipeline's default template: `"This is a photo of {label}."`
2. **Run HF pipeline per image**: `self.pipe(image, candidate_labels=prompts)`. The output will contain the cosine similarity between the image and each label.
3. **Top-1 predicted label** = candidate with highest cosine similarity.
4. **Accuracy** = fraction of samples where predicted label == dataset ground-truth `fine_label`. We report top-1 and top-5 via `TopKAccuracyMetric`.

```
  class names ─▶ prompts ─┐
                          ▼
  image ─────────▶ HF pipeline ─▶ top-1 label ─▶ compare ─▶ accuracy
                                                    ▲
                                                    │
                                            ground truth
```

Models evaluated (CIFAR-100 test, 1000 samples, NPU). "Reported" numbers come from the OpenAI CLIP paper and CLIP_benchmark / open_clip results — our baselines track them to within single-template variance, which confirms the methodology matches the standard.

| Model | HF baseline (fp32, CPU) | ONNX quantized (NPU) | Δ | Reported |
| --- | ---: | ---: | ---: | ---: |
| openai/clip-vit-base-patch32           | 58.4 | 60.7 | +2.3 | ~60 |
| openai/clip-vit-base-patch16           | 62.6 | 63.4 | +0.8 | ~67 |
| openai/clip-vit-large-patch14          | 73.6 | 73.7 | +0.1 | ~77 |
| openai/clip-vit-large-patch14-336      | 73.4 | 73.1 | −0.3 | ~78 |
| laion/CLIP-ViT-B-32-laion2B-s34B-b79K  | 73.0 | 69.3 | −3.7 | ~75 |
| laion/CLIP-ViT-H-14-laion2B-s32B-b79K  | 78.7 | 78.3 | −0.4 | ~82 |
| google/siglip-base-patch16-224         | 68.8 | 66.6 | −2.2 | ~73 |
| google/siglip-so400m-patch14-384       | 80.9 | 79.7 | −1.2 | ~81 |

All NPU-vs-baseline deltas are within normal quantization-noise territory (±~4pp).

## 5. Prompt template

OpenAI's paper curates **different prompt ensembles per dataset** (e.g. 18 templates for CIFAR-100, `"a photo of a {}, a type of food."` for Food101, 48 action templates for UCF101). Ensembling recovers 3-7pp on absolute accuracy.

We do not ship this. Our evaluator is for **regression detection**, not paper reproduction — identical preprocessing between the fp32 and quantized runs is what matters, and a single default template (`"This is a photo of {}."`) satisfies that.

## 6. Key files in this changelist

| File | Description |
| --- | --- |
| `zero_shot_image_classification.py` | New composite model class that orchestrates the split `image-encoder` + `text-encoder` ONNX sessions and produces `ZeroShotImageClassifierOutput` for the HF pipeline. |
| `zero_shot_image_classification_evaluator.py` | New evaluator that delegates per-sample inference to the HF `ZeroShotImageClassificationPipeline` and reports top-1 / top-5 accuracy. |
| `siglip.py` | New HF export config for SigLIP, registering the split text / vision sub-encoders with Optimum. |
| `eval.py` | CLI updated to accept composite models via repeated `-m role=path` pairs, with role-name validation against the composite class's `_SUB_MODEL_CONFIG`. |

### CODE REVIEW

File: src/winml/modelkit/models/winml/composite_model.py [hunk 1]
```diff
@@ -221,6 +221,20 @@ def from_onnx(
         else:
             resolved_cls = cls
 
+        # Validate provided sub-model roles against the resolved class's schema.
```
Comment (vortex-captain) [inline src/winml/modelkit/models/winml/composite_model.py]: A few lines below we have a check similar to the `unknown` check here:
```python
            component_task = resolved_cls._SUB_MODEL_CONFIG.get(name)
            if component_task is None:
                valid = list(resolved_cls._SUB_MODEL_CONFIG.keys())
                raise ValueError(
                    f"Unknown component {name!r}. Valid names for {resolved_cls.__name__}: {valid}"
                )
```
Maybe we could deduplicate a bit here

File: scripts/e2e_eval/run_eval.py [hunk 2]
```diff
@@ -669,7 +669,7 @@ def _run_winml_eval(
     timeout: int,
     ds_config: dict,
     model_dir: Path,
-    onnx_path: str | None = None,
+    onnx_path: dict[str, str] | None = None,
```
Comment (vortex-captain) [inline scripts/e2e_eval/run_eval.py]: nit: `onnx_paths`

File: src/winml/modelkit/eval/metrics/top_k_accuracy.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    Typical usage::
+
+        metric = TopKAccuracyMetric()
+        for pred_labels, gt_label in samples:
+            metric.update(pred_labels, gt_label)
+        result = metric.compute()
+        # {"top1_accuracy": 91.0, "top5_accuracy": 99.0}
+    """
+
+    def __init__(self) -> None:
```
Comment (vortex-captain) [inline src/winml/modelkit/eval/metrics/top_k_accuracy.py]: Curious if it would be useful to make the K in TopK configurable (and possible multiple Ks just in case user need more than one)

File: src/winml/modelkit/models/winml/zero_shot_image_classification.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+# Sub-encoder ONNX output names we accept, in priority order. The first name
+# present in the session outputs wins. CLIP's WithProjection classes expose
+# ``image_embeds`` / ``text_embeds`` natively; SigLIP (no WithProjection
+# variants) exposes the projected embedding under ``pooler_output``.
+_EMBED_OUTPUT_KEYS: tuple[str, ...] = ("image_embeds", "text_embeds", "pooler_output")
+
+
+@register_composite_model("clip", "zero-shot-image-classification")
+@register_composite_model("siglip", "zero-shot-image-classification")
+class WinMLModelForZeroShotImageClassification(WinMLCompositeModel):
```
Comment (vortex-captain) [inline src/winml/modelkit/models/winml/zero_shot_image_classification.py:46]: nit: (question) Exsiting composite model subclasses are named `WinML<TaskOrArch>Model`, would it be necessary to keep same convention?

File: src/winml/modelkit/models/winml/zero_shot_image_classification.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            slice_inputs = {k: v[i:i + 1] for k, v in inputs.items()}
+            all_embeds.append(self._pick_embeds(text._session.run(slice_inputs)))
+        return torch.from_numpy(np.concatenate(all_embeds, axis=0))
+
+    @staticmethod
+    def _pad_or_truncate(arr: np.ndarray, target_len: int) -> np.ndarray:
+        if arr.shape[1] < target_len:
+            pad_width = target_len - arr.shape[1]
+            return np.pad(arr, ((0, 0), (0, pad_width)), constant_values=0)
+        if arr.shape[1] > target_len:
```
Comment (vortex-captain) [inline src/winml/modelkit/models/winml/zero_shot_image_classification.py:174]: nit: when text is truncated, maybe log a warning to user?

File: src/winml/modelkit/models/winml/zero_shot_image_classification.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        inputs = text._format_inputs(**raw)
+        expected = text.io_config["input_shapes"][0][-1]
+        return {k: self._pad_or_truncate(v, expected).astype(np.int64) for k, v in inputs.items()}
+
+    def _run_text(self, inputs: dict[str, np.ndarray]) -> torch.Tensor:
+        """Run text encoder N times (batch=1 ONNX constraint) and concatenate."""
+        text = self.sub_models["text-encoder"]
+        n = inputs["input_ids"].shape[0]
+        all_embeds = []
+        for i in range(n):
```
Comment (vortex-captain) [inline src/winml/modelkit/models/winml/zero_shot_image_classification.py]: we could generalize a bit here as the batch size may not always be 1, simply by running `batch_size` samples every time. Beware the last batch may need padding

File: src/winml/modelkit/models/winml/zero_shot_image_classification.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+            image_embeds=image_embeds,
+        )
+
+    def _preprocess_vision(self, pixel_values: torch.Tensor | None) -> dict[str, np.ndarray]:
+        """Torch→numpy via the sub-model's formatter."""
+        return self.sub_models["image-encoder"]._format_inputs(pixel_values=pixel_values)
+
+    def _run_vision(self, inputs: dict[str, np.ndarray]) -> torch.Tensor:
+        """Run vision encoder once (batch=1 ONNX)."""
+        outputs = self.sub_models["image-encoder"]._session.run(inputs)
```
Comment (vortex-captain) [inline src/winml/modelkit/models/winml/zero_shot_image_classification.py]: maybe consider generalizing to any batch size

**Review by vortex-captain — APPROVED** [review]

**Review by vortex-captain — APPROVED** [review]
